program-runtime/src/compute_budget.rs

```diff
@@ -170,153 +143,24 @@ impl ComputeBudget {
-                || bytes > MAX_HEAP_FRAME_BYTES
-                || bytes < MIN_HEAP_FRAME_BYTES as u32
-                || bytes % 1024 != 0
-            {
-                return Err(TransactionError::InstructionError(
-                    i,
-                    InstructionError::InvalidInstructionData,
-                ));
-            }
-            self.heap_size = Some(bytes as usize);
-        }
-
-        let compute_unit_limit = updated_compute_unit_limit
-            .unwrap_or_else(|| {
-                num_non_compute_budget_instructions
-                    .saturating_mul(DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT)
-            })
-            .min(MAX_COMPUTE_UNIT_LIMIT);
-        self.compute_unit_limit = u64::from(compute_unit_limit);
-
-        self.loaded_accounts_data_size_limit = updated_loaded_accounts_data_size_limit
-            .unwrap_or(MAX_LOADED_ACCOUNTS_DATA_SIZE_BYTES)
-            .min(MAX_LOADED_ACCOUNTS_DATA_SIZE_BYTES);
-
-        Ok(prioritization_fee
-            .map(|fee_type| PrioritizationFeeDetails::new(fee_type, self.compute_unit_limit))
-            .unwrap_or_default())
-    }
-
-    pub fn fee_budget_limits<'a>(
-        instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
-        feature_set: &FeatureSet,
-        maybe_cluster_type: Option<ClusterType>,
-    ) -> FeeBudgetLimits {
-        let mut compute_budget = Self::default();
-
-        // A cluster specific feature gate, when not activated it keeps v1.13 behavior in mainnet-beta;
-        // once activated for v1.14+, it allows compute_budget::request_heap_frame and
-        // compute_budget::set_compute_unit_price co-exist in same transaction.
-        let enable_request_heap_frame_ix = feature_set
-            .is_active(&enable_request_heap_frame_ix::id())
-            || maybe_cluster_type
-                .and_then(|cluster_type| (cluster_type != ClusterType::MainnetBeta).then_some(0))
-                .is_some();
-        let prioritization_fee_details = compute_budget
-            .process_instructions(
-                instructions,
-                !feature_set.is_active(&remove_deprecated_request_unit_ix::id()),
-                enable_request_heap_frame_ix,
-                feature_set.is_active(&add_set_tx_loaded_accounts_data_size_instruction::id()),
-            )
-            .unwrap_or_default();
-
-        FeeBudgetLimits {
-            loaded_accounts_data_size_limit: compute_budget.loaded_accounts_data_size_limit,
-            heap_cost: compute_budget.heap_cost,
-            compute_unit_limit: compute_budget.compute_unit_limit,
-            prioritization_fee: prioritization_fee_details.get_fee(),
-        }
+    pub fn new_from_transaction_meta(transaction_meta: &TransactionMeta) -> Self {
+        let mut compute_budget = ComputeBudget::new(u64::from(transaction_meta.compute_unit_limit));
+        compute_budget.heap_size = transaction_meta.updated_heap_bytes;
+        compute_budget
     }
```

Comment:
improved `ComputeBudget` to be simply a struct has bunch of limits to be passed to runtime, nothing more.